vignettes/piggyback.Rmd

```diff
@@ -36,113 +33,202 @@ number of files or bandwidth to deliver them.
+  read.csv()
+#> # A data.frame: 32 × 12
+#>   X        mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear
+#>   <chr>  <dbl> <int> <dbl> <int> <dbl> <dbl> <dbl> <int> <int> <int>
+#> 1 Mazda…  21       6   160   110  3.9   2.62  16.5     0     1     4
+#> 2 Mazda…  21       6   160   110  3.9   2.88  17.0     0     1     4
+#> 3 Datsu…  22.8     4   108    93  3.85  2.32  18.6     1     1     4
+#> 4 Horne…  21.4     6   258   110  3.08  3.22  19.4     1     0     3
+#> 5 Horne…  18.7     8   360   175  3.15  3.44  17.0     0     0     3
+#> # ℹ 27 more rows
+#> # ℹ 1 more variable: carb <int>
+#> # ℹ Use `print(n = ...)` to see more rows
 ```
+
+Some functions also accept connections when wrapped by `url()`, e.g. for `readRDS()`:
+```{r}
+pb_url <- pb_download_url("mtcars.rds", repo = "tanho63/piggyback-tests", tag = "v0.0.2") %>%
+  url()
+readRDS(pb_url)
+#> # A data.frame: 32 × 11
+#>     mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb
+#>   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
+#> 1  21       6   160   110  3.9   2.62  16.5     0     1     4     4
+#> 2  21       6   160   110  3.9   2.88  17.0     0     1     4     4
+#> 3  22.8     4   108    93  3.85  2.32  18.6     1     1     4     1
+#> 4  21.4     6   258   110  3.08  3.22  19.4     1     0     3     1
+#> 5  18.7     8   360   175  3.15  3.44  17.0     0     0     3     2
+#> # ℹ 27 more rows
+#> # ℹ Use `print(n = ...)` to see more rows
+close(pb_url)
 ```
-[1] "https://github.com/cboettig/piggyback-tests/releases/download/v0.0.1/diamonds.tsv.gz"
-[2] "https://github.com/cboettig/piggyback-tests/releases/download/v0.0.1/iris.tsv.gz"    
-[3] "https://github.com/cboettig/piggyback-tests/releases/download/v0.0.1/iris.tsv.xz"    
-[4] "https://github.com/cboettig/piggyback-tests/releases/download/v0.0.1/iris2.tsv.gz" 
+Note that using `url()` requires that we close the connection after reading it, 
+or else we will receive warnings about leaving open connections.
+
+We can also use this `url()` approach to pass along authentication for private 
+repositories, e.g.
+```{r}
+pb_url <- pb_download_url("mtcars.rds", repo = "tanho63/piggyback-private", url_type = "api") %>%
+  url(
+    headers = c(
+      "Accept" = "application/octet-stream",
+      "Authorization" = paste("Bearer", gh::gh_token())
+    )
+  )
+readRDS(pb_url)
+#> # A tibble: 32 × 11
+#>     mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb
+#>   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
+#> 1  21       6   160   110  3.9   2.62  16.5     0     1     4     4
+#> 2  21       6   160   110  3.9   2.88  17.0     0     1     4     4
+#> 3  22.8     4   108    93  3.85  2.32  18.6     1     1     4     1
+#> 4  21.4     6   258   110  3.08  3.22  19.4     1     0     3     1
+#> 5  18.7     8   360   175  3.15  3.44  17.0     0     0     3     2
+#> # ℹ 27 more rows
+#> # ℹ Use `print(n = ...)` to see more rows
+close(pb_url)
```

Comment:
@cboettig started writing out this "how to download from URL" section and I feel like this got really complicated really quickly. This makes me think we should have a pb_read_url but, unlike with just downloading it to disk, URL support is more complicated especially if we want to abstract the authentication for them:

- base R stuff tends to accept plain URLs and `url()` except `readRDS` which only works with `url()`
- jsonlite supports a url() connection or plain URL
- arrow doesn't accept a `url()` connection but instead wants the plain URL or a raw vector, so we would need to curl it first to pass the auth headers

```r
guess_read_url_function <- function(file) {
  switch(
    file_extension,
    "csv" = function(url, ..., headers) {con<-url(url, headers = headers); on.exit(close(con)); read.csv(con,...)},
    "rds" = function(url, ..., headers) {con<-url(url, headers = headers); on.exit(close(con)); readRDS(con,...)},
    "parquet" = function(url, ..., headers) {httr::GET(url, headers = headers)$content %>% read_parquet(...)},
    stop("unknown fmt")
)
```

Reply:
two paths-ish forward: 

A) tell users to just go to `pb_read` for simplicity if the repo is private
B) make `pb_read_url` happen

I'm ambivalent between the two and unsure how much difference in performance there actually is between pb_read and hypothetical pb_read_url paths...actually I can test that:

Reply:
alright this surprises the heck out of me:

```r
# A tibble: 5 × 13                                                                                   
#> expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time 
#> <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> 
#>   1 pb_read       193ms    206ms      4.90  265.87KB     0        5     0      1.02s 
#>   2 read.csv      603ms    618ms      1.57  147.84KB     0        5     0      3.18s 
#>   3 read_csv      427ms    454ms      2.25    2.09MB     1.50     3     2      1.34s 
#>   4 fread         140ms    143ms      6.50   33.86KB     0        5     0   769.03ms 
#>   5 arrow         604ms    620ms      1.57    2.02MB     0        5     0      3.18s 
```

Reply:
Attempting with larger size...

```r
options(piggyback.verbose = FALSE)
bench::mark(
  pb_read = pb_read("play_by_play_2023.csv", repo = "nflverse/nflverse-data", tag = "pbp"),
  read.csv = read.csv(pb_download_url("play_by_play_2023.csv", repo = "nflverse/nflverse-data", tag = "pbp")),
  read_csv = readr::read_csv(pb_download_url("play_by_play_2023.csv", repo = "nflverse/nflverse-data", tag = "pbp")),
  fread = data.table::fread(pb_download_url("play_by_play_2023.csv", repo = "nflverse/nflverse-data", tag = "pbp")),
  check = FALSE,
  min_iterations = 2
)
#> # A tibble: 4 × 13                                                               
#> expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#> <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
#>   1 pb_read       7.51s    7.62s     0.131     583MB    0.787     2    12     15.24s
#> 2 read.csv      5.52s    5.97s     0.168     582MB    0.838     2    10     11.93s
#> 3 read_csv      4.54s    5.64s     0.177     200MB    0.266     2     3     11.28s
#> 4 fread         3.99s    4.13s     0.242     215MB    0.484     2     4      8.27s
```
okay, phew, must have been overhead or something

Reply:
Opted for option 1 so that I don't have to implement pb_read_url right away. This PR got away from me again...